devito/mpi/distributed.py

```diff
@@ -600,17 +616,53 @@ class CustomTopology(tuple):
     """
 
     def __new__(cls, items, input_comm):
+        # Keep track of nstars and already defined decompositions
         nstars = len([i for i in items if i == '*'])
-        if nstars > 0:
-            if input_comm.size % nstars != 0:
-                raise ValueError("Invalid `topology` for given nprocs")
-            if any(i not in ('*', 1) for i in items):
-                raise ValueError("Custom topology must be only 1 or *")
-
-            v = input_comm.size // nstars
-            processed = [i if i == 1 else v for i in items]
-        else:
+
+        # If no stars exist we are ready
+        if nstars == 0:
             processed = items
+        else:
+            # Init decomposition list
+            processed = [1] * len(items)
+
+            # Get star and integer indices
+            int_pos = [i for i, item in enumerate(items) if isinstance(item, int)]
+            int_vals = [item for item in items if isinstance(item, int)]
+            star_pos = [i for i, item in enumerate(items) if not isinstance(item, int)]
+
+            # Decompose the processes remaining for allocation to prime factors
+            alloc_procs = np.prod([i for i in items if i != '*'])
+            remprocs = int(input_comm.size // alloc_procs)
+            prime_factors = primefactors(remprocs)
```

Comment:
This looks a bit overly intricate. 

If you use `factorint(remproc)` you get directly the dict of factors that you just need to split i.e 

```
factors = factorint(remprocs)
vals = [k for (k, v) in factors.items() for _ in range(v)]
vals = vals + [1 for _ in range(nstars - len(vals))]
startvals = (*vals[:nstart-1], prod(vals[nstars-1:])
```

and should be it

Reply:
Well, not realy, since you do not prioritise neither the outermost dimension, nor the overall balance:
e.g. your approach will give:
```
FAILED tests/test_mpi.py::TestDistributor::test_custom_topology_3d_dummy[24-topology20-dist_topology20] - assert (2, 2, 6) == (6, 2, 2)
FAILED tests/test_mpi.py::TestDistributor::test_custom_topology_3d_dummy[32-topology21-dist_topology21] - assert (2, 2, 8) == (4, 4, 2)
```

Reply:
That's just a minor change  to the last line to cycle instead of multiply the remainder
```
split =  np.array_split(np.pad(vals, nstar)
starvals = np.prod([np.pad(s, (0, nstar-len(s)), constant_values=1) for s in split], dims=1)
```

Reply:
And pad `(nstar-len(s), 0)` if you want it left heavy instead

Reply:
I think I pushed a simplified version, lemme know if you like it better.
I would dare to say that I consider your approach less readable, at least for my eyes.
btw, I could not make it work?
Would you mind to try locally your code?